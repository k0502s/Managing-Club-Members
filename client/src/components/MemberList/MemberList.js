import React, { useState, useEffect } from 'react';
import { Button } from 'reactstrap';
import { Link } from 'react-router-dom';
import Pagination from '@material-ui/lab/Pagination';
import { useSelector, useDispatch } from 'react-redux';
import WarnButton from './Section/WarnButton';
import { MEMBER_DELETE_REQUEST, MEMBER_LIST_REQUEST } from '../../redux/types';

const sex = {
    1: '남',
    2: '여',
};

const MemberList = (props) => {
    const dispatch = useDispatch();
    const [currentTutorial, setCurrentTutorial] = useState(null);
    const [currentIndex, setCurrentIndex] = useState(-1);
    const [searchTitle, setSearchTitle] = useState([]);

    ////페이지 번호/////

    const [page, setPage] = useState(1);
    const [pageSize, setPageSize] = useState(3);
    const { memberlist, totalPages } = useSelector((state) => state.member);
    const pageSizes = [3, 6];

    //현재 상태의 검색창에 입력한 title 값, page, pagesize을 params 객체에 담아준다.
    const getRequestParams = (searchTitle, page, pageSize) => {
        let params = {};

        if (searchTitle) {
            params.title = searchTitle;
        }

        if (page) {
            params.page = page - 1;
        }

        if (pageSize) {
            params.size = pageSize;
        }

        return params;
    };

    //현재 상태 값을 담은 params 객체 데이터를 서버로 보내주어 처리한 데이터를 다시 받아옴
    //받아온 데이터를 현재 상태  state의 리스트 값, 페이지의 conut 값으로 업데이트 해줌
    const retrieveTutorials = () => {
        const params = getRequestParams(searchTitle, page, pageSize);

        dispatch({
            type: MEMBER_LIST_REQUEST,
            payload: { params },
        });
    };

    //retrieveTutorials 변수 선언 아래에 와야함
    useEffect(retrieveTutorials, [page, pageSize]);

    //이벤트 발생하면 받아온 값을 통해 state의 page 값 업데이트
    const handlePageChange = (event, value) => {
        setPage(value);
    };

    //이벤트 발생하여 받아온 값을 보여주는 state의 페이지 갯수를 정해주는 pagesize 값을 업데이트 해주고
    //동시에 페이지 번호를 위한 state의 page 값 초기화 시켜줌.
    const handlePageSizeChange = (event) => {
        setPageSize(event.target.value);
        setPage(1);
    };

    /////리스트 목록//////

    const onChangeSearchTitle = (e) => {
        const searchTitle = e.target.value;
        setSearchTitle(searchTitle);
    };

    const refreshList = () => {
        retrieveTutorials();
        setCurrentTutorial(null);
        setCurrentIndex(-1);
    };

    const refreshList2 = () => {
        retrieveTutorials();
        setCurrentTutorial();
        setCurrentIndex();
    };

    const setActiveTutorial = (tutorial, index) => {
        setCurrentTutorial(tutorial);
        setCurrentIndex(index);
    };

    const deleteTutorial = () => {
        if (currentTutorial) {
            // Axios.delete(`/api/member/${currentTutorial._id}`)
            dispatch({
                type: MEMBER_DELETE_REQUEST,
                payload: currentTutorial._id,
            });
            refreshList2();
        } else {
            refreshList();
        }
    };

    const AddMember = () => {
        props.history.push('/addmember');
    };

    const renderImage = (images) => {
        if (images === []) {
            return `${process.env.REACT_APP_BASIC_IMAGE_URL}`;
        }
        if (images.length > 0) {
            let image = images[0];
            return `${image}`;
            // http://localhost:5000/
        }
    };

    return (
        <div className="list row">
            <div className="col-md-8">
                <div className="input-group mb-3">
                    <input type="text" className="form-control" placeholder="회원 이름을 입력해주세요" value={searchTitle} onChange={onChangeSearchTitle} />
                    <div className="input-group-append">
                        <button className="btn btn-outline-secondary" type="button" onClick={retrieveTutorials}>
                            Search
                        </button>
                    </div>
                </div>
            </div>
            <div className="col-md-6">
                <h4>MEMBER LIST</h4>

                <ul className="list-group">
                    {memberlist &&
                        memberlist.map((memberlist, index) => (
                            <li className={'list-group-item ' + (index === currentIndex ? 'active' : '')} onClick={() => setActiveTutorial(memberlist, index)} key={index}>
                                <h4>{memberlist.title}</h4>
                                <h8>{memberlist.description}</h8>
                            </li>
                        ))}
                </ul>
                <Button className="m-3 btn btn-sm btn-success" onClick={AddMember}>
                    회원 추가
                </Button>
                <Button className="m-3 btn btn-sm btn-danger" onClick={deleteTutorial}>
                    회원 탈퇴
                </Button>
            </div>
            <div className="col-md-6">
                {currentTutorial ? (
                    <div>
                        <h4>MEMBER DATA</h4>
                        <div>
                            <img style={{ width: '130px' }} alt="member" oneEror="this.style.display='none'" src={renderImage(currentTutorial.images)} />
                        </div>
                        <div>
                            <label>
                                <strong>이름:</strong>
                            </label>{' '}
                            {currentTutorial.title}
                        </div>
                        <div>
                            <label>
                                <strong>카메라 기종:</strong>
                            </label>{' '}
                            {currentTutorial.description}
                        </div>
                        <div>
                            <label>
                                <strong>나이 :</strong>
                            </label>{' '}
                            {currentTutorial.price}
                        </div>
                        <div>
                            <label>
                                <strong>성별 :</strong>
                            </label>{' '}
                            {sex[currentTutorial.continents]}
                        </div>
                        <div style={{ display: 'flex' }}>
                            <Link to={'/edit/' + currentTutorial._id} className="m-3 btn-sm btn-success">
                                EDIT
                            </Link>
                            {/* WarnInfo */}
                            <WarnButton detail={currentTutorial} />
                        </div>
                    </div>
                ) : (
                    <div>
                        <br />
                        <p>Please Click on a List...</p>
                    </div>
                )}
            </div>
            <div className="mt-3">
                <h7 style={{ marginLeft: 250 }}>Page: </h7>
                <select onChange={handlePageSizeChange} value={pageSize}>
                    {pageSizes.map((size) => (
                        <option key={size} value={size}>
                            {size}
                        </option>
                    ))}
                </select>

                <Pagination className="my-3" color="primary" count={totalPages} page={page} siblingCount={1} boundaryCount={1} shape="rounded" onChange={handlePageChange} />
            </div>
        </div>
    );
};

export default MemberList;
