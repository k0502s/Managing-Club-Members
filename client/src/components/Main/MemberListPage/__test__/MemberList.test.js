import React from 'react';
import MemberList from '../MemberListPage';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

describe('<MemberList />', () => {
    const mockStore = configureStore();
    let store = mockStore({
        member: {
            memberlist: [
                {
                    name: '김진석',
                },
            ],
        },
    });

    it('matches snapshot', () => {
        const utils = render(
            <Provider store={store}>
                <MemberList />
            </Provider>
        );
        expect(utils.container).toMatchSnapshot();
    });
});
