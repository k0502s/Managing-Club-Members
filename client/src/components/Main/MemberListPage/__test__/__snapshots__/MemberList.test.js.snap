// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MemberList /> matches snapshot 1`] = `
<div>
  <div
    class="list row"
  >
    <div
      class="col-md-8"
    >
      <div
        class="input-group mb-3"
      >
        <input
          class="form-control"
          placeholder="회원 이름을 입력해주세요"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-outline-secondary"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="col-md-6"
    >
      <h4>
        MEMBER LIST
      </h4>
      <ul
        class="list-group"
      >
        <li
          class="list-group-item "
        >
          <h4>
            김진석
          </h4>
          <h8 />
        </li>
      </ul>
      <button
        class="m-3 btn btn-sm btn-success btn btn-secondary"
        type="button"
      >
        회원 추가
      </button>
      <button
        class="m-3 btn btn-sm btn-danger btn btn-secondary"
        type="button"
      >
        회원 탈퇴
      </button>
    </div>
    <div
      class="col-md-6"
    >
      <div>
        <br />
        <p>
          Please Click on a List...
        </p>
      </div>
    </div>
    <div
      class="mt-3"
    >
      <h7
        style="margin-left: 250px;"
      >
        Page: 
      </h7>
      <select>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="6"
        >
          6
        </option>
      </select>
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root my-3"
      >
        <ul
          class="MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-selected"
              tabindex="0"
              type="button"
            >
              1
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded MuiPaginationItem-textPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</div>
`;
